Fix WebGL tests & Implement WebGLRenderingContext::{validateProgram, getProgramInfoLog, disableVertexAttribArray}.

MortimerGoro opened the pull request:
<!-- Please describe your changes on the following line: -->

Implement WebGLRenderingContext::{validateProgram, getProgramInfoLog} and improve WebGL testcase to show shader link and validation errors.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13199 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13208)

<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @larsbergstrom (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webglprogram.rs
 * @emilio: components/script/dom/webglrenderingcontext.rs, components/script/dom/webglprogram.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13198) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1293,6 +1293,11 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
 
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.9
+    fn GetProgramInfoLog(&self, program: Option<&WebGLProgram>) -> Option<DOMString> {
+        program.and_then(|p| Some(p.get_info_log())).map(DOMString::from)
```
nit: This can be `program.map(|p| p.get_info_log()).map(DOMString::from)`

emilio commented:
@bors-servo: r+

Thanks!

bors-servo commented:
:pushpin: Commit 600ac93 has been approved by `emilio`

<!-- @bors-servo r=emilio 600ac93f59fcc1f22e1e37b7d7c4e5ea20433c04 -->

nox commented:
@bors-servo p=1

bors-servo commented:
:hourglass: Testing commit 600ac93 with merge c83e3a4...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3051)

larsbergstrom commented:
Hrm, I've never seen this:
```
command timed out: 1200 seconds without output running ['./mach', 'test-wpt', '--release', '--processes', '8', '--log-raw', 'test-wpt.log', '--log-errorsummary', 'wpt-errorsummary.log'], attempting to kill
```

jdm commented:
@bors-servo: retry
* #13219

bors-servo commented:
:hourglass: Testing commit 600ac93 with merge cc85d61...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3065)

jdm commented:
Huh, it happened in this PR again. I am becoming suspicious...

jdm commented:
This panic also appeared:
```
 ▶ CRASH [expected ERROR] /webgl/conformance-1.0.3/conformance/uniforms/uniform-default-values.html
  │ 
  │ DESCRIPTION: WebGL uniform default values
  │ 
  │ testing: float
  │ *** Error compiling FRAGMENT_SHADER '[object WebGLShader]':ERROR: 0:11: A call to a user defined function is not permitted.
  │ 
  │ FAIL Error in program linking:
  │ Unexpected WebGL error: 0x501 (1281) (thread WebGLThread, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/master/webrender_traits/src/webgl.rs:349)
  │ stack backtrace:
  │    0:        0x1048d605e - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x1048d634c - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x10312be84 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x104963f73 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x104963e44 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:        0x104963d62 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:        0x1046a4f91 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h4fb8cad70c419c4b
  │    7:        0x1046c31c9 - std::panicking::try::do_call::h4ce48cdbe48c6cc2
  │    8:        0x104964b4a - __rust_maybe_catch_panic
  │    9:        0x1046cc283 - _<F as alloc..boxed..FnBox<A>>::call_box::he1289c011bddc36d
  │   10:        0x104962e94 - std::sys::thread::Thread::new::thread_start::h5b631f48cd23f128
  │   11:     0x7fff99f21059 - _pthread_body
  │   12:     0x7fff99f20fd6 - _pthread_start
  │ ERROR:servo: Unexpected WebGL error: 0x501 (1281)
  └ Pipeline failed in hard-fail mode.  Crashing!
```

jdm commented:
There are a bunch of other test failures, too.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13230) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
I think that the problem is related to the webrender version. I only made changes in webrender_tratis which should not affect those tests. I'll wait for #13256

emilio commented:
I don't think we run tests with WebRender right now, though, that's #12989.


On 09/13/2016 05:36 AM, Imanol Fernandez wrote:
> I think that the problem is related to the webrender version. I only
> made changes in webrender_tratis which should not affect those tests.
> I'll wait for #13256 <https://github.com/servo/servo/pull/13256>
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/13208#issuecomment-246667289>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQwutxQ5KwKU2nNByAgjt5x1U-8BrNeks5qppjigaJpZM4J4bl0>.
>

anholt commented:
It looks like `uniform-default-values.html` was exiting early due `getProgramInfoLog` being undefined, so we're probably just proceeding on to something else in the test that generates the error and crashes.  A backtrace with information on which `WebGLCommand` we were processing would be interesting.

If it's not related to `validateProgram`/`getProgramInfoLog`, I think the answer is to `./mach update-wpt` to update the expected output of the tests.  Hopefully you mostly find that a bunch of previously expected failures are no longer present.

MortimerGoro commented:
It worked just rebasing onto master ;)

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 0670776 with merge 5f5a8f7...

highfive commented:
▶ TIMEOUT [expected FAIL] /css-flexbox-1_dev/html/flexbox-basic-iframe-vert-001.htm
      │ 
      └ Shutting down the Constellation after generating an output file or exit flag specified

jdm commented:
That's #13037, and there were no other failures! \o/

emilio commented:
Awesome!

@bors-servo: r+

bors-servo commented:
:pushpin: Commit 0670776 has been approved by `emilio`

<!-- @bors-servo r=emilio 067077647adb5c43b0495b0a479195deb57cfb36 -->

KiChjang commented:
@bors-servo try- clean

KiChjang commented:
@bors-servo r=emilio

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #13172

bors-servo commented:
:pushpin: Commit 0670776 has been approved by `emilio`

<!-- @bors-servo r=emilio 067077647adb5c43b0495b0a479195deb57cfb36 -->

KiChjang commented:
@bors-servo try- r- clean

KiChjang commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 0670776 has been approved by `emilio`

<!-- @bors-servo r=emilio 067077647adb5c43b0495b0a479195deb57cfb36 -->

KiChjang commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 0670776 with merge b821431...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3226)

jdm commented:
@bors-servo: retry
* #13219

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3036), [arm64](http://build.servo.org/builders/arm64/builds/3052), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3105), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3278), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3108), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3103), [windows-dev](http://build.servo.org/builders/windows-dev/builds/998) are reusable. Rebuilding only [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3226)...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3230)

emilio commented:
@bors-servo: retry

 * https://github.com/servo/servo/issues/13219 again

bors-servo commented:
:hourglass: Testing commit 0670776 with merge 944ca23...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3233)

KiChjang commented:
@bors-servo retry
* #13219

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3042), [arm64](http://build.servo.org/builders/arm64/builds/3058), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3111), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3285), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3113), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3109), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1004) are reusable. Rebuilding only [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3233)...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3238)

KiChjang commented:
```
Ran 7571 tests finished in 2732.0 seconds.
  • 7554 ran as expected. 1343 tests skipped.
  • 4 tests crashed unexpectedly
  • 10 tests unexpectedly okay
  • 12 tests had unexpected subtest results

Tests with unexpected results:
  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/conformance/webGLArrays.html

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/drawArraysOutOfBounds.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/drawArraysOutOfBounds.html:
  │ FAIL [expected PASS] WebGL test #0: testDrawArraysEmpty
  │   → assert_true: testDrawArraysEmpty expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/drawArrays.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/drawArrays.html:
  │ FAIL [expected PASS] WebGL test #0: testDrawArraysVBO
  │   → assert_true: testDrawArraysVBO expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/drawElementsBadArgs.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/drawElementsBadArgs.html:
  │ FAIL [expected PASS] WebGL test #0: testDrawElementsVBO
  │   → assert_true: testDrawElementsVBO expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/drawElements.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/drawElements.html:
  │ FAIL [expected PASS] WebGL test #0: testDrawElementsVBO
  │   → assert_true: testDrawElementsVBO expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/texImage2DHTML.html:
  └ PASS [expected FAIL] WebGL test #0: testTexImage2D

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/texSubImage2DHTML.html:
  └ PASS [expected FAIL] WebGL test #0: testTexImage2D

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/vertexAttrib.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/vertexAttrib.html:
  │ FAIL [expected PASS] WebGL test #0: testVertexAttrib
  │   → assert_true: testVertexAttrib expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/vertexAttribBadArgs.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/vertexAttribBadArgs.html:
  │ FAIL [expected PASS] WebGL test #0: testVertexAttrib
  │   → assert_true: testVertexAttrib expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/vertexAttribPointer.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/vertexAttribPointer.html:
  │ FAIL [expected PASS] WebGL test #0: testVertexAttribPointerVBO
  │   → assert_true: testVertexAttribPointerVBO expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:901:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:5
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:189:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:923:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:948:7

  ▶ CRASH [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/more/functions/vertexAttribPointerBadArgs.html
  │ 
  │ Unexpected WebGL error: 0x501 (1281) (thread WebGLThread, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/master/webrender_traits/src/webgl.rs:361)
  │ VertexAttribPointer: Data Type not supported (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }, at /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/webglrenderingcontext.rs:2017)
  │ stack backtrace:
  │    0:        0x10a419cbe - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x10a419fac - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x108ff9204 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x10a8e1eb3 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x10a8e1d84 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:        0x10a8e1ca2 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:        0x10a626aa4 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h484c6905232f80ea
  │    7:        0x10a644e69 - std::panicking::try::do_call::h7834f05248fce601
  │    8:        0x10a8e2a8a - __rust_maybe_catch_panic
  │    9:        0x10a64c613 - _<F as alloc..boxed..FnBox<A>>::call_box::h97958610c9bffe8b
  │   10:        0x10a8e0dd4 - std::sys::thread::Thread::new::thread_start::h022e3887023c6290
  │   11:     0x7fff9406f059 - _pthread_body
  │   12:     0x7fff9406efd6 - _pthread_start
  │ ERROR:servo: Unexpected WebGL error: 0x501 (1281)
  │ stack backtrace:
  │    0:        0x10a419cbe - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x10a419fac - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x108ff9204 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x10a8e1eb3 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x1093db1e4 - std::panicking::begin_panic::h0914615a412ba184
  │    5:        0x109955182 - _<script..dom..webglrenderingcontext..WebGLRenderingContext as script..dom..bindings..codegen..Bindings..WebGLRenderingContextBinding..WebGLRenderingContextBinding..WebGLRenderingContextMethods>::VertexAttribPointer::h211823dc0f226176
  │    6:        0x1094b4c6c - std::panicking::try::do_call::ha3eb7de156b3cda6
  │    7:        0x10a8e2a8a - __rust_maybe_catch_panic
  │    8:        0x1097a1a87 - script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::vertexAttribPointer::h445ebc75c52fe7ad
  │    9:        0x109aa8448 - CallJitMethodOp
  │   10:        0x10967a9b4 - script::dom::bindings::utils::generic_call::h7fc847546e37e590
  │   11:        0x109f0e56e - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │   12:        0x109f0ea25 - 2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_E
  │   13:        0x109e3e5b0 - 2js9fun_applyEP9JSContextjPN2JS5Value
  │   14:        0x109f0e56e - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │   15:        0x109bbf304 - 2js3jitL14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_15ICCall_FallbackEjPN2JS5ValueENS7_13MutableHandleIS8_E
  │ ERROR:servo: VertexAttribPointer: Data Type not supported
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/glsl/arrayOutOfBounds.html:
  └ PASS [expected FAIL] WebGL test #0: testOk

  ▶ OK [expected ERROR] /webgl/conformance-1.0.3/conformance/ogles/GL/ceil/ceil_001_to_006.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/ogles/GL/ceil/ceil_001_to_006.html:
  │ FAIL [expected PASS] WebGL test #0: Error in program linking:
  │   → assert_true: Error in program linking: expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:153:5
  │ linkProgram@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1404:5
  │ createProgram@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1799:3
  │ addShader@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/ogles/ogles-utils.js:576:23
  │ loadShader/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/ogles/ogles-utils.js:552:7
  └ loadTextFileAsync/request.onreadystatechange@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1442:9

  ▶ CRASH [expected ERROR] /webgl/conformance-1.0.3/conformance/ogles/GL/equal/equal_001_to_008.html
  │ 
  │ DESCRIPTION: WebGL GLSL conformance test: equal_001_to_008.html
  │ 
  │ test: equal_vec2_frag.test.html
  │ loading: ../default/default.vert
  │ loading: equal_vec2_frag_ref.frag
  │ loading: ../default/default.vert
  │ loading: equal_vec2_frag.frag
  │ loaded: equal_vec2_frag_ref.frag
  │ *** Error compiling FRAGMENT_SHADER '[object WebGLShader]':ERROR: 0:48: A call to a user defined function is not permitted.
  │ 
  │ loaded: ../default/default.vert
  │ 
  │ FAIL Error in program linking:
  │ Unexpected WebGL error: 0x501 (1281) (thread WebGLThread, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/master/webrender_traits/src/webgl.rs:361)
  │ stack backtrace:
  │    0:        0x10ffd5cbe - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x10ffd5fac - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x10ebb5204 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x11049deb3 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x11049dd84 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:        0x11049dca2 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:        0x1101e2aa4 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h484c6905232f80ea
  │    7:        0x110200e69 - std::panicking::try::do_call::h7834f05248fce601
  │    8:        0x11049ea8a - __rust_maybe_catch_panic
  │    9:        0x110208613 - _<F as alloc..boxed..FnBox<A>>::call_box::h97958610c9bffe8b
  │   10:        0x11049cdd4 - std::sys::thread::Thread::new::thread_start::h022e3887023c6290
  │   11:     0x7fff9406f059 - _pthread_body
  │   12:     0x7fff9406efd6 - _pthread_start
  │ ERROR:servo: Unexpected WebGL error: 0x501 (1281)
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected ERROR] /webgl/conformance-1.0.3/conformance/ogles/GL/floor/floor_001_to_006.html
  │ 
  │ DESCRIPTION: WebGL GLSL conformance test: floor_001_to_006.html
  │ 
  │ test: floor_float_frag_xvary.test.html
  │ loading: ../default/default.vert
  │ loading: floor_float_frag_xvary_ref.frag
  │ loading: ../default/default.vert
  │ loading: floor_float_frag_xvary.frag
  │ loaded: floor_float_frag_xvary_ref.frag
  │ *** Error compiling FRAGMENT_SHADER '[object WebGLShader]':ERROR: 0:43: A call to a user defined function is not permitted.
  │ 
  │ loaded: ../default/default.vert
  │ 
  │ FAIL Error in program linking:
  │ Unexpected WebGL error: 0x501 (1281) (thread WebGLThread, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/master/webrender_traits/src/webgl.rs:361)
  │ stack backtrace:
  │    0:        0x10e58ccbe - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x10e58cfac - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x10d16c204 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x10ea54eb3 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x10ea54d84 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:        0x10ea54ca2 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:        0x10e799aa4 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h484c6905232f80ea
  │    7:        0x10e7b7e69 - std::panicking::try::do_call::h7834f05248fce601
  │    8:        0x10ea55a8a - __rust_maybe_catch_panic
  │    9:        0x10e7bf613 - _<F as alloc..boxed..FnBox<A>>::call_box::h97958610c9bffe8b
  │   10:        0x10ea53dd4 - std::sys::thread::Thread::new::thread_start::h022e3887023c6290
  │   11:     0x7fff9406f059 - _pthread_body
  │   12:     0x7fff9406efd6 - _pthread_start
  │ ERROR:servo: Unexpected WebGL error: 0x501 (1281)
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ OK [expected ERROR] /webgl/conformance-1.0.3/conformance/ogles/GL/notEqual/notEqual_001_to_008.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/ogles/GL/notEqual/notEqual_001_to_008.html:
  │ FAIL [expected PASS] WebGL test #0: Error in program linking:
  │   → assert_true: Error in program linking: expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:153:5
  │ linkProgram@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1404:5
  │ createProgram@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1799:3
  │ addShader@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/ogles/ogles-utils.js:576:23
  │ loadShader/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/ogles/ogles-utils.js:552:7
  └ loadTextFileAsync/request.onreadystatechange@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1442:9

  ▶ CRASH [expected ERROR] /webgl/conformance-1.0.3/conformance/uniforms/uniform-default-values.html
  │ 
  │ DESCRIPTION: WebGL uniform default values
  │ 
  │ testing: float
  │ *** Error compiling FRAGMENT_SHADER '[object WebGLShader]':ERROR: 0:11: A call to a user defined function is not permitted.
  │ 
  │ FAIL Error in program linking:
  │ Unexpected WebGL error: 0x501 (1281) (thread WebGLThread, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/master/webrender_traits/src/webgl.rs:361)
  │ stack backtrace:
  │    0:        0x106b55cbe - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:        0x106b55fac - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:        0x105735204 - servo::main::_{{closure}}::h40c00d857a036dfd
  │    3:        0x10701deb3 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:        0x10701dd84 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:        0x10701dca2 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:        0x106d62aa4 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h484c6905232f80ea
  │    7:        0x106d80e69 - std::panicking::try::do_call::h7834f05248fce601
  │    8:        0x10701ea8a - __rust_maybe_catch_panic
  │    9:        0x106d88613 - _<F as alloc..boxed..FnBox<A>>::call_box::h97958610c9bffe8b
  │   10:        0x10701cdd4 - std::sys::thread::Thread::new::thread_start::h022e3887023c6290
  │   11:     0x7fff9406f059 - _pthread_body
  │   12:     0x7fff9406efd6 - _pthread_start
  │ ERROR:servo: Unexpected WebGL error: 0x501 (1281)
  └ Pipeline failed in hard-fail mode.  Crashing!
```

MortimerGoro commented:
Hi,

I've been working on this issue. A lot of the WebGL conformance tests were exiting early before this PR because getProgramInfoLog function was undefined. After implementing the function that doesn't happen, so more tests are executed, and that requires both fixing the new test expectations and fixing the previous ones.

One of the reasons that getProgramInfoLog func was called so often is that a bunch of shaders were failing to compile. So the tests were exiting early because the unexpected shaders compilation errors and the undefined getProgramInfoLog function. I have submited this PR to fix the shader compilation issues: https://github.com/servo/angle/pull/16 We have to wait for the angle PR, to merge this.

I have fixed all the new tests and updated the expectations. I also improved some of the WebGL code to better fulfill the spec and pass more tests correctly. 

Still a lot of work to do to pass the full webgl conformance ;)

jdm commented on tests/wpt/web-platform-tests/webgl/conformance-1.0.3/conformance/attribs/gl-bindAttribLocation-matrix.html:
```diff
@@ -82,7 +82,11 @@
     gl.attachShader(glProgram, glFragmentShader);
     gl.linkProgram(glProgram);
-    return gl.getProgramParameter(glProgram, gl.LINK_STATUS);
+    var linked =  gl.getProgramParameter(glProgram, gl.LINK_STATUS);
```
This change should be reverted.

emilio commented on components/script/dom/webglprogram.rs:
```diff
@@ -250,6 +293,25 @@ impl WebGLProgram {
+                _ => false
+            };
+            if !shaders_compiled {
+                return Ok("One or more shaders failed to compile".to_string());
```
Hm... Can you specify a bit more what's the error you're trying to avoid with this? `glGetProgramInfoLog` shouldn't generate an error in this case.

MortimerGoro replied:
Before linking a program getProgramInfoLog must return an empty string. After linking a program, if  GL_LINK_STATUS is false, getProgramInfoLog must return a string containing the error description.

We can't rely on the GPU getProgramInfoLog in all the scenarios because in some of them the real glLinkProgram function is never called, and getProgramInfoLog would return always empty string. For example if the WebGLProgram.link() function detects that a shader did not compile succesfully because Angle validation failed.

This error returns the expected message in the scenarios were the real glLinkProgram function has not been called

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -265,6 +265,9 @@ impl WebGLRenderingContext {
         if indx > self.limits.max_vertex_attribs {
             return self.webgl_error(InvalidValue);
         }
+        if indx > self.limits.max_vertex_attribs {
```
duplicated?

MortimerGoro replied:
Fixed

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1270,7 +1279,7 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
     fn DrawElements(&self, mode: u32, count: i32, type_: u32, offset: i64) {
         // From the GLES 2.0.25 spec, page 21:
         //
-        //     "type must be one of UNSIGNED_BYTE or UNSIGNED_SHORT"
```
Reason for this?

MortimerGoro replied:
Reverted, just thought that the extra spaces were a typo.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1281,15 +1290,26 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
         }
 
-        if self.current_program.get().is_none() || self.bound_buffer_element_array.get().is_none() {
+        if self.current_program.get().is_none() {
```
This is also wrong, from the notes section of https://www.khronos.org/opengles/sdk/docs/man/xhtml/glDrawElements.xml :
> If the current program object, as set by glUseProgram, is invalid, rendering results are undefined. However, no error is generated for this case.

So a quote to the spec, and maybe just return here in the `current_program` or `bound_buffer_element_array` are none? (And probably do the same on overflow?)

Probably worth checking what other implementors or drivers do.

MortimerGoro replied:
I'll pass the tests again without checking current_program & bound_buffer_element_array. Just think that some webgl conformance tests expect a error in that cases, but I want to to confirm.

When there is an overflow we must generate the error for sure. The are a lot of tests in the webgl conformance about overflows.

emilio replied:
Ok, please point to those from comments in the code so it's easier to review, and also to check in the future :)

emilio replied:
Also, for the record, I'm not suggesting "don't check the program", but just return early without generating an error, since at least when there's no bound program that's what the spec says.

MortimerGoro replied:
I have found this in WebGL spec: 

> If the CURRENT_PROGRAM is null, an INVALID_OPERATION error will be generated. WebGL performs additional error checking beyond that specified in OpenGL ES 2.0 during calls to drawArrays and drawElements.

Also required to pass some WebGL conformance tests.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1323,25 +1343,46 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
-                None
-            },
-        })
+        if let Some(program) = program {
```
I think here (and below too) will be cleaner like:

```rust
let program = match program {
    Some(program) => program,
    None => {
        self.webgl_error(InvalidValue);
        return None;
     }
};

// And so on.
```

That being said, I can't see what backs on the spec generating an `InvalidValue` error if there's no program, can you find a source?

MortimerGoro replied:
The spec is not clear about an empty program. But the webgl conformance tests seems to expect this error. Example of some of the tests (get-active-test.html):

```
shouldBeNull("context.getActiveUniform(program, 1)");
wtu.glErrorShouldBe(context, context.INVALID_VALUE);
shouldBeNull("context.getActiveUniform(program, -1)");
wtu.glErrorShouldBe(context, context.INVALID_VALUE);
shouldBeNull("context.getActiveUniform(null, 0)");
wtu.glErrorShouldBe(context, context.INVALID_VALUE);
```

MortimerGoro replied:
mmm, we can infer this from the spec: 
> INVALID_VALUE if The value of index is greater than or equal to the number of active uniforms in program.

Any index number will always be greater than the active uniforms in a null program, so it's an invalid value

emilio replied:
Ok, fair enough, can you write that in the code, and point to the `get-active-test.html` file?

emilio reviewed (commented):
I need to properly review the `is_deleted` changes, since it might get tricky because IIRC if a program is attached, it might not be deleted until it's not part of the context's state. In that case, it's necessary to differentiate before requested delete vs. deleted.

This can also be an useful distinction for other objects. That being said, probably it's worth landing those bits assuming the `is_deleted` function returns the correct thing for programs, then implement that properly in another PR.

MortimerGoro commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -644,6 +647,12 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
             constants::NO_ERROR
+            //Todo(Mortimer): Enable GPU getError() when the webtraits repo is merged
+            //let (sender, receiver) = ipc::channel().unwrap();
+            //self.ipc_renderer
```
Ok, this is not required in this PR.

I thought that was interesting to fallback to glGetError so devs can check unexpected errors in problematic GPUs or errors we don't handle yet. I'll check how the asserts are done in the renderer thread. Just curious about performance, are the checks done in release builds too or they should be enabled with a command line option?

emilio replied:
Right now they *are* run, which is a nice perf impact, but I plan to make them not run once https://github.com/servo/servo/pull/13387 is landed.

emilio commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertexattribpointer.html.ini:
```diff
@@ -1,3 +1,3 @@
 [gl-vertexattribpointer.html]
   type: testharness
-  expected: CRASH
+  expected: OK
```
Oh, BTW, the `expected: OK` sections can go away now. I think they're the default.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2015,14 +2061,36 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
         if attrib_id > self.limits.max_vertex_attribs {
             return self.webgl_error(InvalidValue);
         }
+        if stride < 0 || stride > 255 || offset < 0 {
```
please, back this with quotes to the spec, such as:

> From GLES 2.0: GL_INVALID_VALUE is generated if stride is negative.
> From WebGL: If offset is negative, an INVALID_VALUE error will be generated. [...] In WebGL, the maximum supported stride is 255.

Ideally always, but specially when the checks belong to different specs.

emilio commented on components/script/dom/webglshader.rs:
```diff
@@ -114,15 +117,28 @@ impl WebGLShader {
                     self.renderer.send(CanvasMsg::WebGL(msg)).unwrap();
-                    self.compilation_status.set(ShaderCompilationStatus::Succeeded);
+
+                    match self.parameter(constants::COMPILE_STATUS) {
```
We should be pretty confident in that if the shader validation succeeds, the compilation is also successful for the context we're targeting. Do you have an example of something that fails in this case? If that's the case, that's either an ANGLE bug, or that we're using the wrong kind of GL context.

MortimerGoro replied:
It's difficult for me to be confident about this after dealing in the past with some crappy android GPUs :)

Tests pass correctly without this. I haven't found an example that passes validation but fails to compile. I had an issue with this only when loading an OpenGL ES 3.0 context while doing some tests in osx, but the angle validator was set to 2.0

I don't want to prolong this PR, I restored original validation code. We can move this to another PR if needed.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1323,25 +1331,36 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+            None => {
+                self.webgl_error(InvalidValue);
+                return None;
+             }
```
nit: spacing is off by one here.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1323,25 +1331,36 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+            None => {
+                self.webgl_error(InvalidValue);
+                return None;
+             }
```
nit: and here.

emilio commented:
And also, a huge thanks for all the work you're doing here :-)

emilio commented on components/script/dom/webglshader.rs:
```diff
@@ -163,6 +163,12 @@ impl WebGLShader {
         self.info_log.borrow().clone()
     }
 
+    pub fn get_gpu_info_log(&self) -> WebGLResult<String> {
```
This now is unused.

emilio commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertexattribpointer.html.ini:
```diff
@@ -1,3 +1,2 @@
 [gl-vertexattribpointer.html]
```
I think you need to remove the whole section. The usual way of doing this is running `./mach update-wpt /path/to/log` IIRC. I'll run this through automation so you get a log from our builders.

emilio commented:
@bors-servo: try

This is really close, thanks for all this work @MortimerGoro!

bors-servo commented:
:hourglass: Trying commit f4c286a with merge 02c8087...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3288), [arm64](http://build.servo.org/builders/arm64/builds/3301), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3376), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3556), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3365), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3345), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/44), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/44), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1251)

emilio commented:
Could you sqash the first, third and last commit? r=me with that.

So excited to see this landing! :)

MortimerGoro committed:
Fix WebGL tests & Implement WebGLRenderingContext::{validateProgram, getProgramInfoLog, disableVertexAttribArray}, r=emilio

MortimerGoro committed:
Improve test_webgl_triangle.html

MortimerGoro commented:
Done!

emilio commented:
@bors-servo: r+

Yay!

bors-servo commented:
:pushpin: Commit 328dace has been approved by `emilio`

<!-- @bors-servo r=emilio 328dacee8b9abaee40909d11f6c4cf97ecce3870 -->

bors-servo commented:
:hourglass: Testing commit 328dace with merge f9dfb5c...

larsbergstrom commented:
@bors-servo force clean retry
- infra

bors-servo commented:
:hourglass: Testing commit 328dace with merge 2432b48...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1)

larsbergstrom commented:
@bors-servo retry
- infra

bors-servo commented:
:hourglass: Testing commit 328dace with merge 89d85ac...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3386)

MortimerGoro commented:
Could you try again?


Linux:
>    Compiling stylo_tests v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/tests/unit/stylo)
> /home/servo/buildbot/slave/linux-dev/build/tests/unit/stylo/sanity_checks.rs:65:5: 65:84 error: **couldn't read "/home/servo/buildbot/slave/linux-dev/build/tests/unit/stylo/../../../components/style/generated/gecko_pseudo_element_helper.rs": No such file or directory (os error 2)**
> /home/servo/buildbot/slave/linux-dev/build/tests/unit/stylo/sanity_checks.rs:65     include!("../../../components/style/generated/gecko_pseudo_element_helper.rs");
>                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> error: Could not compile `stylo_tests`.
> 
> To learn more, run the command again with --verbose.
> 

Windows:
> Running `rustc C:\projects\servo\.cargo\registry\src\github.com-1ecc6299db9ec823\servo-skia-0.20130412.24\src\lib.rs --crate-name skia --crate-type lib -g -C metadata=821f6e887353da86 -C extra-filename=-821f6e887353da86 --out-dir C:\projects\servo\target\debug\deps --emit=dep-info,link -L dependency=C:\projects\servo\target\debug\deps --extern gleam=C:\projects\servo\target\debug\deps\libgleam-77c1dc4cf66c55c7.rlib --extern libc=C:\projects\servo\target\debug\deps\liblibc-1417726cb94dbc83.rlib --extern expat_sys=C:\projects\servo\target\debug\deps\libexpat_sys-b221d02da6c4f70a.rlib --extern euclid=C:\projects\servo\target\debug\deps\libeuclid-005c17dcf11976c9.rlib --extern glutin=C:\projects\servo\target\debug\deps\libglutin-f1caa1609717624f.rlib --cap-lints allow -W unused-extern-crates -L native=C:\projects\servo\target\debug\build\servo-skia-ba112d30f8f02d37\out/lib -l static=skia -l usp10 -l ole32 -L native=C:\projects\servo\target\debug\build\expat-sys-a8e6de2e5e1a3bbc\out\lib`
> error: failed to build archive: **bad archive: permission denied**
> Build failed, waiting for other jobs to finish...
> error: Could not compile `servo-skia`.

larsbergstrom commented:
@bors-servo retry
- infra

bors-servo commented:
:hourglass: Testing commit 328dace with merge fc25138...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/58)

emilio commented:
@bors-servo: retry

 * https://github.com/servo/servo/issues/13535

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3302), [arm64](http://build.servo.org/builders/arm64/builds/3315), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3392), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3379), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3359), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/60), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1267) are reusable. Rebuilding only [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/58)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3302), [arm64](http://build.servo.org/builders/arm64/builds/3315), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3392), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3379), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3359), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/60), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/60), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1267)

emilio commented:
@bors-servo: retry

 * https://github.com/servo/servo/issues/13535
